hpat/hiframes/pd_dataframe_ext.py

```diff
@@ -625,38 +625,63 @@ def generic(self, args, kws):
-            left, right, left_on, right_on, how)
+        def _impl(left, right, how='inner', on=None, left_on=None,
+                  right_on=None, left_index=False, right_index=False, sort=False,
+                  suffixes=('_x', '_y'), copy=True, indicator=False, validate=None):
+
+            return hpat.hiframes.api.join_dummy(left, right, new_left_on, new_right_on, how)
+    else:
+        def _impl(left, right, how='inner', on=None, left_on=None,
+                  right_on=None, left_index=False, right_index=False, sort=False,
+                  suffixes=('_x', '_y'), copy=True, indicator=False, validate=None):
+            if on is not None:
+                left_on = right_on = on
+
+            return hpat.hiframes.api.join_dummy(left, right, left_on, right_on, how)
 
     return _impl
 
+
 @overload(pd.merge_asof)
 def merge_asof_overload(left, right, on=None, left_on=None, right_on=None,
-        left_index=False, right_index=False, by=None, left_by=None,
-        right_by=None, suffixes=('_x', '_y'), tolerance=None,
-        allow_exact_matches=True, direction='backward'):
-
-    def _impl(left, right, on=None, left_on=None, right_on=None,
-            left_index=False, right_index=False, by=None, left_by=None,
-            right_by=None, suffixes=('_x', '_y'), tolerance=None,
-            allow_exact_matches=True, direction='backward'):
-        if on is not None:
-            left_on = right_on = on
+                        left_index=False, right_index=False, by=None, left_by=None,
+                        right_by=None, suffixes=('_x', '_y'), tolerance=None,
+                        allow_exact_matches=True, direction='backward'):
+
+    # 'on' can be a list or a string literal, use different overloads for them
+    if on is not None and isinstance(on, types.StringLiteral):
+        # use literal strings value to fetch them as consts from IR later
+        new_right_on = on.literal_value
+        new_left_on = on.literal_value
+
+        def _impl(left, right, on=None, left_on=None, right_on=None,
+                  left_index=False, right_index=False, by=None, left_by=None,
+                  right_by=None, suffixes=('_x', '_y'), tolerance=None,
+                  allow_exact_matches=True, direction='backward'):
+
+            return hpat.hiframes.api.join_dummy(left, right, new_left_on, new_right_on, 'asof')
+    else:
+        def _impl(left, right, on=None, left_on=None, right_on=None,
+                  left_index=False, right_index=False, by=None, left_by=None,
+                  right_by=None, suffixes=('_x', '_y'), tolerance=None,
+                  allow_exact_matches=True, direction='backward'):
+            if on is not None:
+                left_on = right_on = on
 
-        return hpat.hiframes.api.join_dummy(
-            left, right, left_on, right_on, 'asof')
+            return hpat.hiframes.api.join_dummy(left, right, left_on, right_on, 'asof')
 
     return _impl
```

Comment:
I'm not sure but I think these functions (including `merge_overload`) might be implemented easier. I mean, use `if\else` clause to calculate proper `left_on` and `right_on` variables and return the same `_impl` function.

Reply:
I tried that at first (i.e. assign new_right/left_on to proper values in context of merge_asof_overload() and remove lines 677 and 678 at all, hence using same body for _impl()), but for some reason it breaks currently working test_join1_seq_key_change1 (the only test we have currently to test "on=['A']" case). In other words, it looks like assignment:
     left_on = right_on = on
only works for lists if it's in the context of _impl(). The error was as below:
     KeyError: "Failed in hpat mode pipeline (step: convert to distributed)\nlist[('A',)](unicode_type)""
probably because the type of 'on' in context of merge_asof_overload() is not the same as type of 'on' in _impl(), hence providing one instead of the other causes failure.
So I ended up using two different overloads - legacy one for lists and new one taking literal values for strings.
Maybe there's a way to copy Numba's ConstList value 'literally' to new_left/right_on variables, but I don't know what it is.